src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,151 @@
+#include <functional>
+#include <utility>
+
+#include "resolve_type.hpp"
+#include "storage/chunk.hpp"
+#include "storage/fixed_width_attribute_vector.hpp"
+#include "storage/reference_segment.hpp"
+#include "storage/table.hpp"
+#include "table_scan.hpp"
+#include "type_cast.hpp"
+
+namespace opossum {
+
+TableScan::TableScan(const std::shared_ptr<const AbstractOperator>& in, const ColumnID column_id,
+                     const ScanType scan_type, const AllTypeVariant search_value)
+    : AbstractOperator(in), _column_id(column_id), _scan_type(scan_type), _search_value(search_value) {}
+
+ColumnID TableScan::column_id() const { return _column_id; }
+
+ScanType TableScan::scan_type() const { return _scan_type; }
+
+const AllTypeVariant& TableScan::search_value() const { return _search_value; }
+
+std::shared_ptr<const Table> TableScan::_on_execute() {
+  auto input_table = _left_input->get_output();
+  const auto output_table = std::make_shared<Table>();
+  const auto column_count = input_table->column_count();
+
+  // If we receive an empty chunk,
+  // we return an empty output table with one chunk only and return early
+  if (input_table->row_count() == 0) {
+    for (ColumnID column_id{0}; column_id < column_count; ++column_id) {
+      output_table->add_column_definition(input_table->column_name(column_id), input_table->column_type(column_id));
+    }
+    return output_table;
+  }
+
+  auto position_list = std::make_shared<PosList>();
```

Comment:
Just a side node: as we parallelize over chunks (not part of the lecture), creating a large single output chunk can have a negative performance impact on the following operators.

Reply:
Also, if you create one reference segment per input segment, you can optimize your implementation of scanning reference segments (by reusing specialized implementations for value/dictionary segments and iterating over the PosList rather than the full segments).